src/lazy/expanded/mod.rs

```diff
@@ -131,84 +134,334 @@ impl<'top, 'data, D: LazyDecoder<'data>> ExpandedStreamItem<'top, 'data, D> {
+                    } else {
+                        // If the expression was equivalent to `(:void)`, return to the top of
+                        // the loop and get the next expression.
+                        continue;
+                    }
                 }
-                EndOfStream => ExpandedStreamItem::EndOfStream,
+                EndOfStream => return Ok(SystemStreamItem::EndOfStream),
             };
-            return Ok(expanded_item);
+        }
+    }
+
+    /// If there is not an evaluation in process, returns `Ok(None)`.
+    /// If there is an evaluation in process but it does not yield another value, returns `Ok(None)`.
+    /// If there is an evaluation in process and it yields another value, returns `Ok(Some(value))`.
+    /// Otherwise, returns `Err`.
+    fn next_from_evaluator<'top>(
+        &'top self,
+    ) -> IonResult<Option<SystemStreamItem<'top, 'data, D>>> {
+        let evaluator_ptr = match self.evaluator_ptr.get() {
+            // There's not currently an evaluator.
+            None => return Ok(None),
+            // There's an evaluator in the process of expanding a macro.
+            Some(ptr) => ptr,
+        };
+        let evaluator = Self::ptr_to_evaluator(evaluator_ptr);
+
+        match evaluator.next(self.context(), 0) {
+            Ok(Some(value)) => {
+                // See if this value was a symbol table that needs interpretation.
+                self.interpret_value(value).map(Some)
+            }
+            Ok(None) => {
+                // While the evaluator had macros in its stack, they did not produce any more
+                // values. The stack is now empty.
+                Ok(None)
+            }
+            Err(e) => Err(e),
         }
     }
 }
 
 /// The source of data backing a [`LazyExpandedValue`].
-#[derive(Debug, Clone)]
+#[derive(Copy, Clone)]
 pub enum ExpandedValueSource<'top, 'data, D: LazyDecoder<'data>> {
     /// This value was a literal in the input stream.
     ValueLiteral(D::Value),
     /// This value was part of a template definition.
-    Template(&'top Element),
-    /// This value was the computed result of a macro invocation like `(:make_string ...)`.
+    Template(Environment<'top, 'data, D>, TemplateElement<'top>),
+    /// This value was the computed result of a macro invocation like `(:make_string `...)`.
     Constructed(
         // TODO: Make this an associated type on the LazyDecoder trait so 1.0 types can set
         //       it to `Never` and the compiler can eliminate this code path where applicable.
-        (
-            // A collection of bump-allocated annotation strings
-            BumpVec<'top, &'top str>,
-            ExpandedValueRef<'top, 'data, D>,
-        ),
+        // Constructed data stored in the bump allocator. Holding references instead of the data
+        // itself allows this type (and those that contain it) to impl `Copy`.
+        &'top [&'top str],
+        &'top ExpandedValueRef<'top, 'data, D>,
```

Comment:
🗺️ These values all live in the bump allocator. Switching them to references allowed `LazyExpandedValue` to implement `Copy`.